### ISSUE TITLE
twisted.protocols.htb.Bucket.drip() returns invalid results

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5448]
|<img alt="gsk's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| gsk reported|
|-|-|
|Trac ID|trac#5448|
|Type|defect|
|Created|2012-01-10 06:08:56Z|
|Branch|https://github.com/twisted/twisted/tree/htb-drip-5448|

From the api docs (http://twistedmatrix.com/documents/current/api/twisted.protocols.htb.Bucket.html)

The method drip() of the Bucket class
"Returns	True if I am now empty."

The current drip() returns True only if the rate is zero (causing the bucket to drain instantaneously). This behavior does not take into account the possibility that a bucket will drain itself over time.

Attached is a patch containing a unit test for this issue and changes to the Bucket class that remedies this issue.

Attachments:

* [bucket.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7f2/7f25e70c2e6d91d4c4a908f9ea301f9a6827f45a/01117e497e104b645bdb25d9d6310fbcb516cd02.patch) (1824 bytes) - added by gsk on 2012-01-10 06:10:12Z - Unit test and patch to Bucket class

<details><summary>Searchable metadata</summary>

```
trac-id__5448 5448
type__defect defect
reporter__gsk gsk
priority__normal normal
milestone__ 
branch__branches_htb_drip_5448 branches/htb-drip-5448
branch_author__gsk gsk
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1326175736000000 1326175736000000
changetime__1326767848000000 1326767848000000
version__None None
owner__jesstess jesstess
cc__jesstess cc__thijs cc__gsk
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_2"></a>|@jesstess commented|
|-|-|

(In [33448]) Branching to 'htb-drip-5448'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess commented|
|-|-|

(In [33449]) Apply bucket.patch by gsk.

refs [#5448](https://github.com/twisted/twisted/issues/5448)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess commented|
|-|-|

Thanks for the bug report and patch, gsk! I've applied your patch with a couple of cosmetic changes so that it conforms to our coding standards. (The code surrounding your changes is quite old and doesn't set a good example :) ) You can see the changes in revision r33450, and read our coding standard [here](http://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html).

[Buildbot results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/htb-drip-5448) look good.

I'll go ahead and merge.

Thanks again for the patch!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess set status to `closed`|
|-|-|

(In [33452]) Merge htb-drip-5448

Author: gsk
Review: jesstess
Fixes: [#5448](https://github.com/twisted/twisted/issues/5448)

Fix twisted.protocols.htb.Bucket.drip to drip and empty properly with
a non-zero drip rate.
